### ISSUE TITLE
mismatch behavior of XCTestScaffold.runAllTests in Xcode and SwiftPM

### ISSUE THREAD

**Issue by Kyle-Ye** [apple/swift-testing#329]
### Description

Suppose we have multi test targets in a swift package, eg. testTarget - A, B, C.

When we convert them into using `swift-testing` framework. We need to add the `Scaffolding.swift` file currently(for Swift 5.10) to each of the test target so that we can run each of the testTarget(A/B/C) in Xcode.

![image](https://github.com/apple/swift-testing/assets/43724855/dcbc0b45-2c4e-4150-8711-257ffd6c41df)

![image](https://github.com/apple/swift-testing/assets/43724855/fbaea686-df26-46dd-9ef1-2476de4ba76a)

If we only add one in testTarget A, we won't be able to run `@Test` defined in B and C.

```
final class AllTests: XCTestCase {
    func testAll() async {
        await XCTestScaffold.runAllTests(hostedBy: self)
    }
}
```

Things looks pretty good so far. The test will pass on Xcode and SwiftPM via `swift test`.

But after I add a test case which can run only once, I found the CI is failing. Checking the log I found the that case is running 3 times.

So for testCase `T` defined in testTarget A, it will be executed 3 times each for A, B, C if we run `swift test` via SwiftPM.

```
Test Case '-[DemoKitTests3.AllTests testAll]' started.
􀟈  Test run started.
􀄵  Testing Library Version: unknown (Swift 5.10 toolchain)
􀟈  Suite T started.
􀟈  Test testDemo() started.
􁁛  Test testDemo() passed after 0.001 seconds.
􁁛  Suite T passed after 0.001 seconds.
􁁛  Test run with 1 test passed after 0.001 seconds.
Test Case '-[DemoKitTests3.AllTests testAll]' passed (0.007 seconds).
Test Suite 'AllTests' passed at 2024-04-05 00:56:57.643.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.007 (0.007) seconds
Test Suite 'AllTests' started at 2024-04-05 00:56:57.643.
Test Case '-[DemoKitTests2.AllTests testAll]' started.
􀟈  Test run started.
􀄵  Testing Library Version: unknown (Swift 5.10 toolchain)
􀟈  Suite T started.
􀟈  Test testDemo() started.
􀢄  Test testDemo() recorded an issue at T.swift:7:9: Expectation failed: (Demo.shared → DemoKit.Demo) == nil
􀢄  Test testDemo() failed after 0.001 seconds with 1 issue.
􀢄  Suite T failed after 0.001 seconds with 1 issue.
􀢄  Test run with 1 test failed after 0.002 seconds with 1 issue.
/Users/kyle/Downloads/DemoKit/Tests/DemoKitTests1/T.swift:7: error: -[DemoKitTests2.AllTests testAll] :
Test Case '-[DemoKitTests2.AllTests testAll]' failed (0.176 seconds).
Test Suite 'AllTests' failed at 2024-04-05 00:56:57.819.
	 Executed 1 test, with 1 failure (0 unexpected) in 0.176 (0.176) seconds
Test Suite 'AllTests' started at 2024-04-05 00:56:57.819.
Test Case '-[DemoKitTests1.AllTests testAll]' started.
􀟈  Test run started.
􀄵  Testing Library Version: unknown (Swift 5.10 toolchain)
􀟈  Suite T started.
􀟈  Test testDemo() started.
􀢄  Test testDemo() recorded an issue at T.swift:7:9: Expectation failed: (Demo.shared → DemoKit.Demo) == nil
􀢄  Test testDemo() failed after 0.001 seconds with 1 issue.
􀢄  Suite T failed after 0.001 seconds with 1 issue.
􀢄  Test run with 1 test failed after 0.001 seconds with 1 issue.
/Users/kyle/Downloads/DemoKit/Tests/DemoKitTests1/T.swift:7: error: -[DemoKitTests1.AllTests testAll] :
Test Case '-[DemoKitTests1.AllTests testAll]' failed (0.002 seconds).
Test Suite 'AllTests' failed at 2024-04-05 00:56:57.821.
	 Executed 1 test, with 1 failure (0 unexpected) in 0.002 (0.002) seconds
Test Suite 'DemoKitPackageTests.xctest' failed at 2024-04-05 00:56:57.821.
	 Executed 3 tests, with 2 failures (0 unexpected) in 0.185 (0.185) seconds
Test Suite 'All tests' failed at 2024-04-05 00:56:57.821.
	 Executed 3 tests, with 2 failures (0 unexpected) in 0.185 (0.186) seconds
```

### Expected behavior

`T` will only executed by testTarget A.

### Actual behavior

`T` will be executed by all testTargets which contains `XCTestScaffold.runAllTests(hostedBy: self)`

### Steps to reproduce

[DemoKit.zip](https://github.com/apple/swift-testing/files/14874597/DemoKit.zip)

### swift-testing version/commit hash

0.6.0

**Comment by Kyle-Ye** [comment]
**Low priority** for this issue since `XCTestScaffold` is already marked deprecated and `swift-testing` does not support `Swift 5.10` officially.

Just report it as an issue in case other guys encounter it too.

**Comment by grynspan** [comment]
We don't officially support `XCTestScaffold`, and there is no integration between swift-testing and Xcode, so this is not something we expect to resolve. If we _did_ look at resolving it, we would need to make changes in the Swift 5.10 toolchain, but that's already been released, and this sort of issue probably doesn't meet the criteria for a 5.10 dot release.

It's expected that you'd add a separate call to `XCTestScaffold.runAllTests(hostedBy:)` to each test target. Or, more accurately, we don't define how things work with `XCTestScaffold` when you have multiple test targets. There's no _great_ workaround in practice because Xcode runs XCTest-based tests using a different mechanism than SwiftPM. Xcode builds each test target as a separate product, while SwiftPM builds them all into a single product.

What you could do is for tests that are sensitive to running multiple times, add some global state that you set when the test starts, and skip the test if it's already set. Something like:

```swift
static var alreadyRan = Atomic<Bool>(rawValue: false)
@Test(.disabled(if: alreadyRan.load())) func f() {
  alreadyRan.store(true)

  ...
}
```

(That's assuming you've got an atomics library handy, and that my example code compiles, mind you.)
